// @mistralai/mistralai/esm/models/components/jsonpatchpayloadresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: aa05b4aa9186
 */
import * as z from "zod/v4";
import { safeParse } from "../../lib/schemas.js";
import { discriminatedUnion } from "../../types/discriminatedUnion.js";
import { JSONPatchAdd$inboundSchema } from "./jsonpatchadd.js";
import { JSONPatchAppend$inboundSchema, } from "./jsonpatchappend.js";
import { JSONPatchRemove$inboundSchema, } from "./jsonpatchremove.js";
import { JSONPatchReplace$inboundSchema, } from "./jsonpatchreplace.js";
/** @internal */
export const JSONPatchPayloadResponseValue$inboundSchema = discriminatedUnion("op", {
    add: JSONPatchAdd$inboundSchema,
    append: JSONPatchAppend$inboundSchema,
    remove: JSONPatchRemove$inboundSchema,
    replace: JSONPatchReplace$inboundSchema,
});
export function jsonPatchPayloadResponseValueFromJSON(jsonString) {
    return safeParse(jsonString, (x) => JSONPatchPayloadResponseValue$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'JSONPatchPayloadResponseValue' from JSON`);
}
/** @internal */
export const JSONPatchPayloadResponse$inboundSchema = z.object({
    type: z.literal("json_patch"),
    value: z.array(discriminatedUnion("op", {
        add: JSONPatchAdd$inboundSchema,
        append: JSONPatchAppend$inboundSchema,
        remove: JSONPatchRemove$inboundSchema,
        replace: JSONPatchReplace$inboundSchema,
    })),
});
export function jsonPatchPayloadResponseFromJSON(jsonString) {
    return safeParse(jsonString, (x) => JSONPatchPayloadResponse$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'JSONPatchPayloadResponse' from JSON`);
}
