// @mistralai/mistralai/esm/models/components/turbinetoollocale.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 4a2e89dd045a
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
/** @internal */
export const TurbineToolLocale$inboundSchema = z.object({
    name: z.nullable(z.record(z.string(), z.string())).optional(),
    description: z.nullable(z.record(z.string(), z.string())).optional(),
    usage_sentence: z.nullable(z.record(z.string(), z.string())).optional(),
    working_description: z.nullable(z.record(z.string(), z.string())).optional(),
    done_description: z.nullable(z.record(z.string(), z.string())).optional(),
}).transform((v) => {
    return remap$(v, {
        "usage_sentence": "usageSentence",
        "working_description": "workingDescription",
        "done_description": "doneDescription",
    });
});
export function turbineToolLocaleFromJSON(jsonString) {
    return safeParse(jsonString, (x) => TurbineToolLocale$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'TurbineToolLocale' from JSON`);
}
