// @mistralai/mistralai/esm/models/components/turbinetoolmeta.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 044cdb287e1c
 */
import * as z from "zod/v4";
import { remap as remap$ } from "../../lib/primitives.js";
import { safeParse } from "../../lib/schemas.js";
import { ToolType$inboundSchema } from "./tooltype.js";
import { TurbineToolLocale$inboundSchema, } from "./turbinetoollocale.js";
/** @internal */
export const TurbineToolMeta$inboundSchema = z.object({
    locale: z.nullable(TurbineToolLocale$inboundSchema).optional(),
    tool_type: z.nullable(ToolType$inboundSchema).optional(),
    timeout: z.nullable(z.number()).optional(),
    private_execution: z.nullable(z.boolean()).optional(),
}).transform((v) => {
    return remap$(v, {
        "tool_type": "toolType",
        "private_execution": "privateExecution",
    });
});
export function turbineToolMetaFromJSON(jsonString) {
    return safeParse(jsonString, (x) => TurbineToolMeta$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'TurbineToolMeta' from JSON`);
}
