// @mistralai/mistralai/esm/models/components/updatedocumentrequest.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 * @generated-id: 56629a04610c
 */
import * as z from "zod/v4";
import { smartUnion } from "../../types/smartUnion.js";
/** @internal */
export const Attributes$outboundSchema = smartUnion([
    z.boolean(),
    z.string(),
    z.int(),
    z.number(),
    z.date().transform(v => v.toISOString()),
    z.array(z.string()),
    z.array(z.int()),
    z.array(z.number()),
    z.array(z.boolean()),
]);
export function attributesToJSON(attributes) {
    return JSON.stringify(Attributes$outboundSchema.parse(attributes));
}
/** @internal */
export const UpdateDocumentRequest$outboundSchema = z.object({
    name: z.nullable(z.string()).optional(),
    attributes: z.nullable(z.record(z.string(), smartUnion([
        z.boolean(),
        z.string(),
        z.int(),
        z.number(),
        z.date().transform(v => v.toISOString()),
        z.array(z.string()),
        z.array(z.int()),
        z.array(z.number()),
        z.array(z.boolean()),
    ]))).optional(),
});
export function updateDocumentRequestToJSON(updateDocumentRequest) {
    return JSON.stringify(UpdateDocumentRequest$outboundSchema.parse(updateDocumentRequest));
}
